openfl-tutorials/interactive_api/PyTorch_Huggingface_transformers_SUPERB/workspace/PyTorch_Huggingface_transformers_SUPERB.ipynb

```diff
@@ -0,0 +1,485 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "foreign-gospel",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "\"\"\"\n",
+    "Download the pretrained model and fine-tune it. For classification we use the AutoModelForAudioClassification class.\n",
+    "\"\"\"\n",
+    "\n",
+    "num_labels = len(id2label)\n",
+    "\n",
+    "model = AutoModelForAudioClassification.from_pretrained(\n",
+    "    model_checkpoint,\n",
+    "    num_labels=num_labels,\n",
+    "    label2id=label2id,\n",
+    "    id2label=id2label,\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "0d5afa68-4bd3-43d8-a86d-d59b5cad94bd",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from transformers import AdamW\n",
+    "\n",
+    "params_to_update = []\n",
+    "for param in model.parameters():\n",
+    "    if param.requires_grad == True:\n",
+    "        params_to_update.append(param)\n",
+    "\n",
+    "optimizer = AdamW(params_to_update, lr=3e-5)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "caroline-passion",
+   "metadata": {},
+   "source": [
+    "#### Register model"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "handled-teens",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "framework_adapter = (\n",
+    "    \"openfl.plugins.frameworks_adapters.pytorch_adapter.FrameworkAdapterPlugin\"\n",
+    ")\n",
+    "MI = ModelInterface(\n",
+    "    model=model, optimizer=optimizer, framework_plugin=framework_adapter\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "portuguese-groove",
+   "metadata": {},
+   "source": [
+    "### Define and register FL tasks"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "5d4ff313-a17f-4119-a4c7-afa898b0f304",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "batch_size = 16\n",
+    "args = TrainingArguments(\n",
+    "    \"finetuned_model\",\n",
+    "    save_strategy=\"epoch\",\n",
+    "    per_device_train_batch_size=batch_size,\n",
+    "    per_device_eval_batch_size=batch_size,\n",
+    "    num_train_epochs=1,\n",
+    "    warmup_ratio=0.1,\n",
+    "    logging_steps=10,\n",
+    "    push_to_hub=False,\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "increasing-builder",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "TI = TaskInterface()\n",
+    "\n",
+    "import torch.nn as nn\n",
+    "import tqdm\n",
+    "\n",
+    "\n",
+    "@TI.register_fl_task(\n",
+    "    model=\"model\", data_loader=\"train_loader\", device=\"device\", optimizer=\"optimizer\"\n",
+    ")\n",
+    "def train(model, train_loader, optimizer, device, some_parameter=None):\n",
+    "\n",
+    "    print(f\"\\n\\n TASK TRAIN GOT DEVICE {device}\\n\\n\")\n",
+    "\n",
+    "    model = nn.DataParallel(model, device_ids=[int(device.split(\":\")[1])])\n",
```

Comment:
That's correct. The problem is that by default, the trainer argument detects all gpus on the machine (10 in our case) and sets n_gpus to that number (https://github.com/huggingface/transformers/blob/57882177becb85560f1ff931abb1b0b75d67e70d/src/transformers/training_args.py#L1060) . This can be seen when the trainer args are printed:

Instantaneous batch size per device = 16
Total train batch size (w.parallel, distributed & accumulation) = 160

This leads to the runtime error: cuda runtime error (711) : peer mapping resources exhausted at ../aten/src/THC/THCGeneral.cpp:139

This error comes when more than 8 peer gpu devices are being used and hence I explicitly set the device. Somehow, CUDA_VISIBILE_DEVICES isn't working. Alternatively, trainer.args._n_gpu = 1 is working, but I am not sure if that is a good practice to set.

Reply:
Okay, that could be left here as a workaround then, however it's strange that `CUDA_VISIBLE_DEVICES` variable is not working